File: LAB1/C#/LAB1_pop/Stopper.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace LAB1_pop
{
    public class Stopper
    {
        private Calculator[] _calculators;

        public Stopper(Calculator[] calculators)
        {
            _calculators = calculators;

            Array.Sort(_calculators);
        }

        public void Stop()
        {
            int currentWaitedTime = 0;

            for (int i = 0; i < _calculators.Length; i++)
            {
                int waitingTime = _calculators[i].WorkTime - currentWaitedTime;
                Thread.Sleep(waitingTime);
                currentWaitedTime += waitingTime;
                _calculators[i].Stop();
            }
        }
    }
}

File: LAB1/Java/src/Stopper.java
import java.lang.reflect.Array;
import java.util.Arrays;

public class Stopper implements Runnable {
    private Calculator[] _calculators;

    public Stopper(Calculator[] calculators){
        _calculators = calculators;
        Arrays.sort(_calculators);
    }

    @Override
    public void run() {
        long currentWaitedTime = 0;

        for (int i = 0; i<_calculators.length;i++){
            long waitingTime = _calculators[i].getWorkTime() - currentWaitedTime;
            try {
                Thread.sleep(waitingTime);
                currentWaitedTime += waitingTime;
                _calculators[i].stopProcess();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }
}
